Feat: search tracks and artists on iTunes

idipanshu opened the pull request:
### Description

- Search tracks and artists using iTunes API
- Result displayed in a grid layout
- Click on grid card to view details of the track

---

### Steps to Reproduce / Test

Run ```yarn test -u```

---

### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

### GIF's
![next1](https://user-images.githubusercontent.com/105773536/171181328-f6e14b00-610e-4fb7-a0de-f6092230763d.gif)

---

idipanshu committed:
Setting up template for iTunes project

- Updated .env files
- Removed unwanted files
- Updated Global styling and colors

idipanshu committed:
feat: added search tracks feature

- Landing page with search tracks option
- Created actions, reducer, selectors for fetching data from iTunes API
- Created new api client for iTunes

idipanshu committed:
feat: Track Details Page

- Track details page when music card from grid is clicked

idipanshu committed:
fix: fixed import errors

christin-wednesday commented on environments/.env.development:
```diff
@@ -1 +1 @@
-NEXT_PUBLIC_GITHUB_URL=https://api.github.com/
\ No newline at end of file
+NEXT_PUBLIC_ITUNES_URL=https://itunes.apple.com/
```
Please add new a line at the end of files

idipanshu replied:
Noted and updated 👍🏻

christin-wednesday commented on next.config.js:
```diff
@@ -11,6 +11,7 @@ const constructAlias = (config) => {
   return {
     '@app': path.resolve(__dirname, './app'),
     '@components': path.resolve(__dirname, './app/components'),
+    '@containers': path.resolve(__dirname, './app/containers'),
```
👍🏽

christin-wednesday commented on pages/index.js:
```diff
@@ -1,24 +1,179 @@
+// export async function getStaticProps() {
```
Please remove the commented out code

idipanshu replied:
Sure

christin-wednesday replied:
+1

christin-wednesday commented on pages/index.js:
```diff
@@ -1,24 +1,179 @@
+  && {
+    padding: 0.1rem;
+    margin: 1rem 0;
+    background-color: #fafafa;
```
Please use predefined variables for colors.

idipanshu replied:
I have updated it to use predefined vars.

christin-wednesday commented on pages/index.js:
```diff
@@ -1,24 +1,179 @@
+  const [loading, setLoading] = useState(false);
+  const [playTrackEvent, setPlayTrackEvent] = useState('');
+
+  const tracks = Object.entries(songsData);
```
please add null check here

idipanshu replied:
Updated 👍🏻

christin-wednesday commented on pages/index.js:
```diff
@@ -1,24 +1,179 @@
-export function ReposPage({ recommendations }) {
-  return <Repos recommendations={recommendations} />;
+    const refs = {};
+    tracks.map((track) => {
```
please add null check

christin-wednesday commented on pages/tracks/[trackId].js:
```diff
@@ -0,0 +1,182 @@
+    `} */
+`;
+
+function SongDetailsPage({ intl, dispatchGetSongDetails, songsData, trackDetails, songsError }) {
```
Let's move this to the details container

idipanshu committed:
Refactoring

- Added new empty line in .env files

christin-wednesday commented on app/components/Header/index.js:
```diff
@@ -0,0 +1,43 @@
+  && {
+    margin-bottom: 0;
+    ${fonts.style.heading};
+    font-family: verdana, sans-serif;
```
font should be defined in the app level.

christin-wednesday commented on app/components/MusicCard/index.js:
```diff
@@ -0,0 +1,161 @@
+      </If>
+
+      <FlexView>
+        <AlbumArt src={artworkUrl100} alt="album art" />
```
Please add check for album art url

christin-wednesday commented on app/components/MusicCard/index.js:
```diff
@@ -0,0 +1,161 @@
+          <Text>
+            {intl.formatMessage({ id: 'track_price' })}:{' '}
+            <Bold>
+              {trackPrice} {currency}
```
We should generally add null checks when rendering properties from server responses.

christin-wednesday commented on app/containers/ItunesTrackFinder/saga.js:
```diff
@@ -0,0 +1,39 @@
+  if (ok) {
+    const result = {};
+
+    data.results.forEach((song) => {
```
please use reduce here

christin-wednesday commented on app/containers/ItunesTrackFinder/selectors.js:
```diff
@@ -0,0 +1,27 @@
+export const selectSearchedTerm = () =>
+  createSelector(selectSearchContainerDomain, (substate) => get(substate, 'searchedTerm'));
+
+export const selectTrackDetails = () =>
```
i recommend constants for these keys

christin-wednesday replied:
please check the existing containers

christin-wednesday commented on pages/index.js:
```diff
@@ -1,24 +1,179 @@
+    setPlayTrackEvent(trackId);
+  }
+
+  const debounceOnChange = debounce(handleOnChange, 500);
```
please move this to `handleOnChange` so that we can remove this extra function definition

christin-wednesday commented on pages/tracks/[trackId].js:
```diff
@@ -0,0 +1,182 @@
+`;
+const Heading = styled.h1`
+  font-size: 1.7rem;
+  font-weight: 600;
```
please use themes for typography properties

christin-wednesday commented on pages/tracks/[trackId].js:
```diff
@@ -0,0 +1,182 @@
+  flex-direction: row;
+  gap: 1rem;
+
+  /* ${media.lessThan('desktop')`
```
please remove

idipanshu replied:
Done 👍🏻

idipanshu committed:
Merge branch 'master' into feat/itunes-tracks-finder

idipanshu committed:
refactor: add null checks

- Added null checks for the data returned from server
- Added new file for selector keys

idipanshu committed:
refactor: Moved container data from pages dir to container dir

- Moved SearchTrack to container dir
- Moved TrackDetails to container dir

idipanshu committed:
refactor/fix: update .eslintrc.js and tests for iTunes project

- Updated .eslintrc.js to add container dir alias
- Updated tests
